### DISCUSSION TITLE
How does npm audit compare to Checkmarx?

### DISCUSSION THREAD

**Discussion by mellis481** [npm/feedback#241] [General]
My company uses a software security solution called Checkmarx to scan products for vulnerabilities.  These scans are done on all areas of the system (eg. database, backend/API, etc.) , not just the UI.  I'm unclear on how it conducts a scan of the UI, but the vulnerabilities it's reporting are... interesting.  For instance, I'm unclear how a dependency of a dependency of a webpack-dev-server (which is only used for development) would be reported as a "high vulnerability".

Running `npm audit fix --force` does not identify any vulnerabilities in my UI.  Upgrading transient dependencies for no good reason is not something I want to do so if my organization can use `npm audit` to satisfactorily scan for vulnerabilities, I will try to advocate for that.

To that end, I'm looking for input on how `npm audit` compares to other security/vulnerability scanning tools.

**Comment by MylesBorins** [comment]
Hey @mellis481. I'm not super familliar with Checkmarx but it seems like some of the products they offer are related to static code analysis, is that what you are using?

npm audit is checking the versions in your dependency tree against our [security advisory database](https://www.npmjs.com/advisories) and informs you if there are known vulnerabilities in any of the packages in your tree. This is not a static analysis tool and relies on exist known and reported vulnerabilities to identify issues.

**Reply by mellis481** [reply]
When you say npm audit isn't a static code analysis tool, you're saying it checks dependency versions for security issues whereas a static code analysis tool looks at the application/production code, correct?

How does npm audit work exactly? Does it scan every dependency listed in the package-lock.json? If so, how does it differentiate between dependencies with vulnerabilities that could actually be exploited vs. dependencies with vulnerabilities that aren't compiled?

**Reply by ljharb** [reply]
It doesn't. Lodash, for example, contains like 30-50 functions. There may be a vulnerability in just one of them, and as such, any usage, direct or transitive, of lodash alerts every consumer of lodash regardless of whether they use the vulnerable part or not.

**Reply by mellis481** [reply]
@ljharb  Thanks. What about dependencies that don't get compiled and are, therefore, not available through the UI to users? Like if webpack uses a dependency that uses a dependency that has a reported vulnerability, does npm audit recognize that and flag it?

**Reply by MylesBorins** [reply]
@mellis481 one thing worth clarifying is that JS is not compiled.

npm audit give you a warning for any vulnerability int he package tree as documented by your package.json / package-lock.json. It does not have any insight into your code or if you are affected by the vulnerability. It is the job of you are your team to determine if you are affected by the vulnerability or not. It is indeed possible that a vulnerability will be identified by audit that does not result in a vulnerability being exposed by your application
